app/controllers/problems_controller.rb

```diff
@@ -87,7 +87,13 @@ def submit
       @problem = Problem.find(params[:id])
       @submission = Submission.new(submit_params) # create submission
       respond_to do |format|
-        if @submission.save
+        source_is_valid = @submission.source.valid_encoding? && !@submission.source.include?("\0")
```

Comment:
Nice work. Just one slight issue – `@submission.source` seems to be `nil` if the user doesn't select a file, which causes this to raise an exception instead of displaying an error message ("Source can't be blank").